eth2-educators/eth2-docker
# Permissions

Permissions work

Local user ID matters for deposit-cli, so user can read created files and deposit-cli can write into directory

For existing beacon/validator containers, use a docker-entrypoint.sh that starts as root, chowns the directory,
then uses gosu (Debian) or su-exec (Alpine) to run as a non-privileged user.

This method of running as root first will be removed again when all users have upgraded, maybe 2 months
after Altair. Warnings in release notes should be given ahead of time.

Testing eth1:
- `sudo bash`
- `chown -R 1000:1000 /var/lib/docker/volumes/VOLUMENAME/*`
- `ls /var/lib/docker/volumes/VOLUMENAME`
- In separate window run client, then verify permissions were changed:
- `ls /var/lib/docker/volumes/VOLUMENAME` 

Tests run
- [x] Geth binary
- [x] Geth source
- [x] OE binary
- [x] OE source
- [x] NM binary
- [x] NM source
- [x] Besu binary
- [x] Besu source

Testing eth2:
- `sudo bash`
- `chown -R 1000:1000 /var/lib/docker/volumes/VOLUMENAME/*`
- `ls /var/lib/docker/volumes/VOLUMENAME`
- In separate window run `docker-compose up -d eth2`, then verify permissions were changed:
- `ls /var/lib/docker/volumes/VOLUMENAME`
- Check logs of beacon and validator and see that they can access their storage
- `docker-compose down` 
- Import keys in separate window, with or without storing passwords, then verify that permissions are right:
- `ls -lR /var/lib/docker/volumes/VOLUMENAME`
- Run eth2 again, verify that key is seen by validator client
- Run command to list accounts, if it exists; verify it can access accounts
- Run command for voluntary exit, if it exists; verify it can access accounts -> This test requires
  a fully synced beacon, which can take days. Current code has been only minimally tested; this
  will be done after the initial rc1
- `docker-compose down` 

- [x] LH binary
- [x] LH source
- [x] Teku binary - validator-exit tested successfully
- [x] Teku source
- [x] Nimbus binary
- [x] Nimbus source
- [x] Prysm binary
- [x] Prysm source

Testing deposit-cli:

- [x] deposit-cli-new
- [x] deposit-cli-new with --uid
- [x] deposit-cli-new with --folder
- [x] deposit-cli-new with both --uid and --folder
- [x] All four for deposit-cli-existing
